### PR TITLE
Fix vertex shader types and tests

### LINKED ISSUE #345: Fix vertex buffer type checking
Make vertex buffers properly check if they are one on the supported types. Currently it only works for vectors and primitives - it should also work for vectors.

### PR DESCRIPTION
**Pull request by reczkok**
closes #345

### CODE REVIEW

**Review by mhawryluk — COMMENTED** [review]
looks good to me! let's add some tests to make sure it works from now on

**Comment by reczkok** [comment]
# TLDR
In the lates commit I also reworked the `unmanagedIO` and `runtime` tests. The previous implementation was less than optimal and made it difficult to write complex tests. Now they should be a little easier to work with 😁

# Details
Previously when creating mocks I used the following method
```ts
const mockCommandEncoder = vi.fn(() => ({
  beginComputePass: vi.fn(() => mockComputePassEncoder()),
  beginRenderPass: vi.fn(),
  ...
}));

const mockDevice = vi.fn(() => ({
  createBindGroup: vi.fn(() => 'mockBindGroup'),
  createBindGroupLayout: vi.fn(() => 'mockBindGroupLayout'),
  createBuffer: vi.fn(() => mockBuffer()),
  createCommandEncoder: vi.fn(() => mockCommandEncoder()),
  ...
});
```
Every mock was essentially a mock creator (i.e. each call to `mockDevice()` (or any `mockSth`) would create a new Mock object. This meant extracting info about calls from nested Mocks was very inconvenient as there were no handles to them.
Now all mocks are objects so we have all the necessary handles. Using `beforeEach()` and `restoreAllMocks()` now has sense and properly resets mocks. All this should mean tests that are easier to understand and expand🫡

**Review by mhawryluk — APPROVED** [review]
Great!!

File: packages/typegpu/tests/runtime.test.ts [hunk 1]
```diff
@@ -257,4 +270,163 @@ describe('TgpuRuntime', () => {
...
+    });
+  });
+
+  it('should throw an error when trying to create an invalid vertex buffer', async () => {
+    const bufferData = tgpu
+      .createBuffer(
+        struct({
+          i: vec2f,
+          should: vec3f,
+          throw: u32,
```
Comment (iwoplaza) [inline packages/typegpu/tests/runtime.test.ts:410]: 💨🏈

**Review by iwoplaza — APPROVED** [review]
Awesome! 🔧 👷
